[#16] Adding Scatter Gather extension queries

hsenasilva opened the pull request:
Hi!
This pull request should resolve #16

Thanks

zambrovski reviewed (changes requested):
Please add `@since 0.2.0` annotation to KDoc of the functions.

hsenasilva commented:
> Please add `@since 0.2.0` annotation to KDoc of the functions.

Hi!
done

hsenasilva commented:
Hi @smcvb, @zambrovski any more changes in this pull request?

smcvb commented:
Although I know the original issue talked about using [Kotlin.time Duration](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/), I feel the mix of this extension being experimental, in combination with the Kotlin's duration to (still) be experimental to make the use case of these methods a little to low.

Hence I'd opt to move to the `scatterGather` default of `long`/`TimeUnit`, or potentially Java's `Duration`, at least for this PR.
I feel that enhances the usability of your PR for the time being. Added, _as_ the Kotlin Extension is also still experimental, we are free to change it to Kotlin's Duration once it no longer is experimental.

So, let us know what you think @hsenasilva. If you agree, please adjust your PR slightly. After that I'll have a further go at reviewing. :-)

hsenasilva commented:
> Although I know the original issue talked about using [Kotlin.time Duration](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-duration/), I feel the mix of this extension being experimental, in combination with the Kotlin's duration to (still) be experimental to make the use case of these methods a little to low.
> 
> Hence I'd opt to move to the `scatterGather` default of `long`/`TimeUnit`, or potentially Java's `Duration`, at least for this PR.
> I feel that enhances the usability of your PR for the time being. Added, _as_ the Kotlin Extension is also still experimental, we are free to change it to Kotlin's Duration once it no longer is experimental.
> 
> So, let us know what you think @hsenasilva. If you agree, please adjust your PR slightly. After that I'll have a further go at reviewing. :-)

Hi @smcvb, 
It was a good point of view, i really understand that ExperimentalTime may be a blocker to someone use these functions.
I changed the PR, could you review? 
Thank you

sandjelkovic commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/QueryGatewayExtensions.kt:
```diff
@@ -106,3 +108,114 @@ inline fun <reified R, reified Q> QueryGateway.queryForOptional(query: Q): Compl
+ * @see ResponseTypes
+ * @since 0.2.0
+ */
+inline fun <reified R, reified Q> QueryGateway.scatterGatherForSingle(queryName: String, query: Q, timeout: Long,
```
The new convention is `scatterGather` without `ForSingle`

sandjelkovic commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/QueryGatewayExtensions.kt:
```diff
@@ -106,3 +108,114 @@ inline fun <reified R, reified Q> QueryGateway.queryForOptional(query: Q): Compl
+ * @see ResponseTypes
+ * @since 0.2.0
+ */
+inline fun <reified R, reified Q> QueryGateway.scatterGatherForMultiple(query: Q, timeout: Long,
```
The new convention is `scatterGatherMany`, without `For` and `Many` instead of `Multiple`

sandjelkovic commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/QueryGatewayExtensions.kt:
```diff
@@ -106,3 +108,114 @@ inline fun <reified R, reified Q> QueryGateway.queryForOptional(query: Q): Compl
+ * @see ResponseTypes
+ * @since 0.2.0
+ */
+inline fun <reified R, reified Q> QueryGateway.scatterGatherForSingle(query: Q, timeout: Long,
```
The new convention is `scatterGather` without `ForSingle`

sandjelkovic commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/QueryGatewayExtensions.kt:
```diff
@@ -106,3 +108,114 @@ inline fun <reified R, reified Q> QueryGateway.queryForOptional(query: Q): Compl
+ * @see ResponseTypes
+ * @since 0.2.0
+ */
+inline fun <reified R, reified Q> QueryGateway.scatterGatherForMultiple(queryName: String, query: Q, timeout: Long,
```
The new convention is `scatterGatherMany`, without `For` and `Many` instead of `Multiple`

sandjelkovic commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/QueryGatewayExtensions.kt:
```diff
@@ -106,3 +108,114 @@ inline fun <reified R, reified Q> QueryGateway.queryForOptional(query: Q): Compl
+ * @see ResponseTypes
+ * @since 0.2.0
+ */
+inline fun <reified R, reified Q> QueryGateway.scatterGatherForOptional(query: Q, timeout: Long,
```
The new convention is `scatterGatherOptional` without `For`

sandjelkovic commented on kotlin/src/main/kotlin/org/axonframework/extensions/kotlin/QueryGatewayExtensions.kt:
```diff
@@ -106,3 +108,114 @@ inline fun <reified R, reified Q> QueryGateway.queryForOptional(query: Q): Compl
+ * @see ResponseTypes
+ * @since 0.2.0
+ */
+inline fun <reified R, reified Q> QueryGateway.scatterGatherForOptional(queryName: String, query: Q, timeout: Long,
```
The new convention is `scatterGatherOptional` without `For`

sandjelkovic reviewed (changes requested):
Hi  @hsenasilva, thanks for the update! Since the original issue is a bit older, I forgot to update its description, both for text and method names. That's my bad.

In the meantime, we changed collection-based methods to conform to `Many` naming instead of `Multiple`, to match the method names in another (reactor) extension. In addition, we removed `For` and `Single` from method names.

I have now updated the original issue with new information, both for Kotlin time and Multiple/Many and Single method names.

So if you could rename methods, that would be perfect. Other than that, the PR looks good!

sandjelkovic reviewed (approved):
Looks good to me 👍

smcvb reviewed (approved):
Looks good to me :+1:

smcvb commented:
Let's ignore the slack authorization issue of the time being...I trust @hsenasilva provided something which works.
And if it doesn't, we'll fix it afterwards.
